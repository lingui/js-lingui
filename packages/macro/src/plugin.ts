import type { PluginObj, Visitor, PluginPass } from "@babel/core"
import type * as babelTypes from "@babel/types"
import MacroJSX from "./macroJsx"
import { NodePath } from "@babel/traverse"
import MacroJs from "./macroJs"
import { MACRO_PACKAGE } from "./constants"
import {
  type LinguiConfigNormalized,
  getConfig as loadConfig,
} from "@lingui/conf"
import { Program, Identifier } from "@babel/types"

let config: LinguiConfigNormalized

export type LinguiPluginOpts = {
  // explicitly set by CLI when running extraction process
  extract?: boolean
  linguiConfig?: LinguiConfigNormalized
}

function getConfig(_config?: LinguiConfigNormalized) {
  if (_config) {
    config = _config
  }
  if (!config) {
    config = loadConfig()
  }
  return config
}

function reportUnsupportedSyntax(path: NodePath, e: Error) {
  const codeFrameError = path.buildCodeFrameError(
    `Unsupported macro usage. Please check the examples at https://lingui.dev/ref/macro#examples-of-js-macros.
 If you think this is a bug, fill in an issue at https://github.com/lingui/js-lingui/issues

 Error: ${e.message}`
  )

  // show stack trace where error originally happened
  codeFrameError.stack = e.stack
  throw codeFrameError
}

type LinguiSymbol = "Trans" | "useLingui" | "i18n"

export default function ({
  types: t,
}: {
  types: typeof babelTypes
}): PluginObj {
  function addImport(state: PluginPass, name: LinguiSymbol) {
    const path = state.get(
      "macroImport"
    ) as NodePath<babelTypes.ImportDeclaration>

    const config = state.get("linguiConfig") as LinguiConfigNormalized

    if (!state.get("has_import_" + name)) {
      state.set("has_import_" + name, true)
      const [moduleSource, importName] = config.runtimeConfigModule[name]

      const [newPath] = path.insertAfter(
        t.importDeclaration(
          [
            t.importSpecifier(
              getSymbolIdentifier(state, name),
              t.identifier(importName)
            ),
          ],
          t.stringLiteral(moduleSource)
        )
      )

      path.parentPath.scope.registerDeclaration(newPath)
    }

    return path.parentPath.scope.getBinding(
      getSymbolIdentifier(state, name).name
    )
  }

  function getMacroImports(path: NodePath<Program>) {
    return path.get("body").filter((statement) => {
      return (
        statement.isImportDeclaration() &&
        statement.get("source").node.value === MACRO_PACKAGE
      )
    })
  }

  function getSymbolIdentifier(
    state: PluginPass,
    name: LinguiSymbol
  ): Identifier {
    return state.get("linguiIdentifiers")[name]
  }

  return {
    name: "lingui-macro-plugin",
    visitor: {
      Program: {
        enter(path, state) {
          const macroImports = getMacroImports(path)

          if (!macroImports.length) {
            return
          }

          state.set("macroImport", macroImports[0])

          state.set(
            "linguiConfig",
            getConfig((state.opts as LinguiPluginOpts).linguiConfig)
          )

          state.set("linguiIdentifiers", {
            i18n: path.scope.generateUidIdentifier("i18n"),
            Trans: path.scope.generateUidIdentifier("Trans"),
            useLingui: path.scope.generateUidIdentifier("useLingui"),
          })

          path.traverse(
            {
              JSXElement(path, state) {
                const macro = new MacroJSX(
                  { types: t },
                  {
                    transImportName: getSymbolIdentifier(state, "Trans").name,
                    stripNonEssentialProps:
                      process.env.NODE_ENV == "production" &&
                      !(state.opts as LinguiPluginOpts).extract,
                  }
                )

                let newNode: false | babelTypes.Node

                try {
                  newNode = macro.replacePath(path)
                } catch (e) {
                  reportUnsupportedSyntax(path, e as Error)
                }

                if (newNode) {
                  const [newPath] = path.replaceWith(newNode)
                  addImport(state, "Trans").reference(newPath)
                }
              },

              "CallExpression|TaggedTemplateExpression"(
                path: NodePath<
                  | babelTypes.CallExpression
                  | babelTypes.TaggedTemplateExpression
                >,
                state: PluginPass
              ) {
                const macro = new MacroJs(
                  { types: t },
                  {
                    stripNonEssentialProps:
                      process.env.NODE_ENV == "production" &&
                      !(state.opts as LinguiPluginOpts).extract,
                    i18nImportName: getSymbolIdentifier(state, "i18n").name,
                    useLinguiImportName: getSymbolIdentifier(state, "useLingui")
                      .name,
                  }
                )
                let newNode: false | babelTypes.Node

                try {
                  newNode = macro.replacePath(path)
                } catch (e) {
                  reportUnsupportedSyntax(path, e as Error)
                }

                if (newNode) {
                  const [newPath] = path.replaceWith(newNode)

                  if (macro.needsUseLinguiImport) {
                    addImport(state, "useLingui").reference(newPath)
                  }

                  if (macro.needsI18nImport) {
                    addImport(state, "i18n").reference(newPath)
                  }
                }
              },
            } as Visitor<PluginPass>,
            state
          )
        },
        exit(path, state) {
          const macroImports = getMacroImports(path)
          macroImports.forEach((path) => path.remove())
        },
      },
    } as Visitor<PluginPass>,
  }
}
