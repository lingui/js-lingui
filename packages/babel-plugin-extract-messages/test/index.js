import fs from "fs"
import path from "path"
import { transformFileSync } from "babel-core"

import plugin from "@lingui/babel-plugin-extract-messages"

const LOCALE_DIR = "./locale"

const rmdir = dir => {
  if (!fs.existsSync(dir)) return
  const list = fs.readdirSync(dir)

  for (let i = 0; i < list.length; i++) {
    const filename = path.join(dir, list[i])
    const stat = fs.statSync(filename)

    if (filename === "." || filename === "..") {
      // pass these files
    } else if (stat.isDirectory()) {
      // rmdir recursively
      rmdir(filename)
    } else {
      // rm fiilename
      fs.unlinkSync(filename)
    }
  }
  fs.rmdirSync(dir)
}

function testCase(testName, assertion) {
  const transform = (filename, jsx = true) => () =>
    transformFileSync(path.join(__dirname, "fixtures", filename), {
      babelrc: false,
      plugins: [
        ...(/integration.*\.js$/.test(filename)
          ? jsx
            ? [
                "@lingui/babel-plugin-transform-js",
                "@lingui/babel-plugin-transform-react"
              ]
            : ["@lingui/babel-plugin-transform-js"]
          : []),
        [
          plugin,
          {
            localeDir: LOCALE_DIR
          }
        ],
        ...(jsx ? ["babel-plugin-syntax-jsx"] : [])
      ]
    })

  it(testName, () => assertion(transform))
}

describe("babel-plugin-lingui-extract-messages", function() {
  // CWD is root directory of repository, so origin of all messages is going to
  // relative to root
  const buildDir = path.join(
    LOCALE_DIR,
    "_build",
    "packages",
    "babel-plugin-extract-messages",
    "test",
    "fixtures"
  )

  beforeAll(() => {
    rmdir(LOCALE_DIR)
  })

  afterAll(() => {
    rmdir(LOCALE_DIR)
  })

  testCase(
    "should raise exception on duplicate id and different defaults",
    transform => {
      expect(transform("jsx/duplicate-id-valid.js")).not.toThrow()
      expect(transform("jsx/duplicate-id.js")).toThrow(/Different defaults/)
    }
  )

  testCase(
    "shouldn't write catalog for files without translatable messages",
    transform => {
      expect(transform("empty.js")).not.toThrow()
      expect(fs.existsSync(path.join(buildDir, "empty.js.json"))).toBeFalsy()
    }
  )

  testCase("should preserve path to file inside locale dir", transform => {
    expect(transform("jsx/deep/all.js")).not.toThrow()
    expect(
      fs.existsSync(path.join(buildDir, "jsx", "deep", "all.js.json"))
    ).toBeTruthy()
  })

  testCase("should ignore files without lingui import", transform => {
    expect(transform("jsx/without-lingui.js")).not.toThrow()
    expect(
      fs.existsSync(path.join(buildDir, "jsx", "without-lingui.js.json"))
    ).toBeFalsy()
  })

  testCase("should extract noop strings", transform => {
    const result = transform("noop/actual.js")()

    const expected = fs.readFileSync(
      path.join(__dirname, "fixtures", "noop", "expected.js")
    )
    expect(result.code.replace(/\r/g, "").trim()).toEqual(
      expected
        .toString()
        .replace(/\r/g, "")
        .trim()
    )

    const messages = JSON.parse(
      fs.readFileSync(path.join(buildDir, "noop", "actual.js.json"))
    )
    expect(messages).toMatchSnapshot()
  })

  testCase("should extract all messages from JSX files", transform => {
    // first run should create all required folders and write messages
    expect(transform("jsx/all.js")).not.toThrow()
    // another runs should just write messages
    expect(transform("jsx/all.js")).not.toThrow()

    const messages = JSON.parse(
      fs.readFileSync(path.join(buildDir, "jsx", "all.js.json"))
    )
    expect(messages).toMatchSnapshot()
  })

  testCase(
    "should extract all messages from JSX files (integration)",
    transform => {
      // first run should create all required folders and write messages
      expect(transform("jsx/integration.js")).not.toThrow()
      // another runs should just write messages
      expect(transform("jsx/integration.js")).not.toThrow()

      const messages = JSON.parse(
        fs.readFileSync(path.join(buildDir, "jsx", "integration.js.json"))
      )
      expect(messages).toMatchSnapshot()
    }
  )

  testCase(
    "should extract all messages from JSX files with macro (integration)",
    transform => {
      // first run should create all required folders and write messages
      expect(transform("jsx/integration-with-macro.js")).not.toThrow()
      // another runs should just write messages
      expect(transform("jsx/integration-with-macro.js")).not.toThrow()

      const messages = JSON.parse(
        fs.readFileSync(
          path.join(buildDir, "jsx", "integration-with-macro.js.json")
        )
      )
      expect(messages).toMatchSnapshot()
    }
  )

  testCase(
    "should extract all messages from JSX files (integration with alises)",
    transform => {
      // first run should create all required folders and write messages
      expect(transform("jsx/integration-with-aliases.js")).not.toThrow()
      // another runs should just write messages
      expect(transform("jsx/integration-with-aliases.js")).not.toThrow()

      const messages = JSON.parse(
        fs.readFileSync(
          path.join(buildDir, "jsx", "integration-with-aliases.js.json")
        )
      )
      expect(messages).toMatchSnapshot()
    }
  )

  testCase("should extract all messages from JS files", transform => {
    // first run should create all required folders and write messages
    expect(transform("js/all.js", false)).not.toThrow()
    // another runs should just write messages
    expect(transform("js/all.js", false)).not.toThrow()

    const messages = JSON.parse(
      fs.readFileSync(path.join(buildDir, "js", "all.js.json"))
    )
    expect(messages).toMatchSnapshot()
  })

  testCase(
    "should extract all messages from JS files (integration)",
    transform => {
      // first run should create all required folders and write messages
      expect(transform("js/integration.js", false)).not.toThrow()
      // another runs should just write messages
      expect(transform("js/integration.js", false)).not.toThrow()

      const messages = JSON.parse(
        fs.readFileSync(path.join(buildDir, "js", "integration.js.json"))
      )
      expect(messages).toMatchSnapshot()
    }
  )

  it("should extract JS translations only once inside React components", function() {
    expect(() =>
      transformFileSync(
        path.join(__dirname, "fixtures", "jsx", "with-react.js"),
        {
          babelrc: false,
          plugins: [
            "@lingui/babel-plugin-transform-js",
            "@lingui/babel-plugin-transform-react",
            [
              plugin,
              {
                localeDir: LOCALE_DIR
              }
            ]
          ],
          presets: ["react"]
        }
      )
    ).not.toThrow()

    const messages = JSON.parse(
      fs.readFileSync(path.join(buildDir, "jsx", "with-react.js.json"))
    )
    expect(messages).toMatchSnapshot()
  })
})
