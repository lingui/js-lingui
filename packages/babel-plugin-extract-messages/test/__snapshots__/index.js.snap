// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`babel-plugin-lingui-extract-messages should extract JS translations only once inside React components 1`] = `
Object {
  Test: Object {
    origin: Array [
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/jsx/with-react.js,
        3,
      ],
    ],
  },
}
`;

exports[`babel-plugin-lingui-extract-messages should extract all messages from JS files (integration) 1`] = `
Object {
  msg.hello: Object {
    origin: Array [
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/js/integration.js,
        1,
      ],
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/js/integration.js,
        2,
      ],
    ],
  },
}
`;

exports[`babel-plugin-lingui-extract-messages should extract all messages from JS files 1`] = `
Object {
  msg.default: Object {
    defaults: Hello World,
    origin: Array [
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/js/all.js,
        5,
      ],
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/js/all.js,
        6,
      ],
    ],
  },
  msg.hello: Object {
    origin: Array [
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/js/all.js,
        4,
      ],
    ],
  },
  {count, plural, one {# book} other {# books}}: Object {
    origin: Array [
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/js/all.js,
        7,
      ],
    ],
  },
}
`;

exports[`babel-plugin-lingui-extract-messages should extract all messages from JSX files (integration with alises) 1`] = `
Object {
  Hi, my name is {name}: Object {
    origin: Array [
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/jsx/integration-with-aliases.js,
        3,
      ],
    ],
  },
}
`;

exports[`babel-plugin-lingui-extract-messages should extract all messages from JSX files (integration) 1`] = `
Object {
  Hi, my name is {name}: Object {
    origin: Array [
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/jsx/integration.js,
        3,
      ],
    ],
  },
  Title: Object {
    origin: Array [
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/jsx/integration.js,
        4,
      ],
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/jsx/integration.js,
        13,
      ],
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/jsx/integration.js,
        14,
      ],
    ],
  },
  {count, plural, one {# book} other {# books}}: Object {
    origin: Array [
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/jsx/integration.js,
        6,
      ],
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/jsx/integration.js,
        16,
      ],
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/jsx/integration.js,
        21,
      ],
    ],
  },
}
`;

exports[`babel-plugin-lingui-extract-messages should extract all messages from JSX files 1`] = `
Object {
  Hi, my name is <0>{name}</0>: Object {
    origin: Array [
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/jsx/all.js,
        9,
      ],
    ],
  },
  msg.default: Object {
    defaults: Hello World,
    origin: Array [
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/jsx/all.js,
        7,
      ],
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/jsx/all.js,
        8,
      ],
    ],
  },
  msg.hello: Object {
    origin: Array [
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/jsx/all.js,
        6,
      ],
    ],
  },
  {count, plural, one {# book} other {# books}}: Object {
    origin: Array [
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/jsx/all.js,
        10,
      ],
    ],
  },
}
`;

exports[`babel-plugin-lingui-extract-messages should extract all messages from JSX files with macro (integration) 1`] = `
Object {
  Title: Object {
    origin: Array [
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/jsx/integration-with-macro.js,
        4,
      ],
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/jsx/integration-with-macro.js,
        13,
      ],
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/jsx/integration-with-macro.js,
        14,
      ],
    ],
  },
  {count, plural, one {# book} other {# books}}: Object {
    origin: Array [
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/jsx/integration-with-macro.js,
        6,
      ],
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/jsx/integration-with-macro.js,
        16,
      ],
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/jsx/integration-with-macro.js,
        21,
      ],
    ],
  },
}
`;

exports[`babel-plugin-lingui-extract-messages should extract noop strings 1`] = `
Object {
  Czech: Object {
    origin: Array [
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/noop/actual.js,
        5,
      ],
    ],
  },
  English: Object {
    origin: Array [
      Array [
        packages/babel-plugin-extract-messages/test/fixtures/noop/actual.js,
        4,
      ],
    ],
  },
}
`;
